File: garbage/solution/garbage-sub_3_st.py
XY_LIMIT = 1000000000

class point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

class node:
    def __init__(self, low, high):
        self.low = low
        self.high = high
        self.val = 0
        self.left = None
        self.right = None

    def insert(self, pos, val):
        self.val += val
        if (self.low == self.high): return

        mid = (self.low + self.high) // 2
        if (pos <= mid):
            if (self.left == None): self.left = node(self.low, mid)
            self.left.insert(pos, val)
        else:
            if (self.right == None): self.right = node(mid+1, self.high)
            self.right.insert(pos, val)

    def sum(self, l, h):
        if (h < self.low or l > self.high): return 0
        if (l <= self.low and h >= self.high): return self.val

        res = 0
        if (self.left != None): res += self.left.sum(l, h)
        if (self.right != None): res += self.right.sum(l, h)
        return res

class node2d:
    def __init__(self, low, high):
        self.low = low
        self.high = high
        self.tree = node(0, XY_LIMIT-1)
        self.left = None
        self.right = None

    def insert(self, x, y, val):
        self.tree.insert(y, val)
        if (self.low == self.high): return

        mid = (self.low + self.high) // 2
        if (x <= mid):
            if (self.left == None): self.left = node2d(self.low, mid)
            self.left.insert(x, y, val)
        else:
            if (self.right == None): self.right = node2d(mid+1, self.high)
            self.right.insert(x, y, val)

    def sum(self, l, r, b, t):
        if (r < self.low or l > self.high): return 0
        if (l <= self.low and r >= self.high): return self.tree.sum(b, t)

        res = 0
        if (self.left != None): res += self.left.sum(l, r, b, t)
        if (self.right != None): res += self.right.sum(l, r, b, t)
        return res

# read input
N, W, H = map(int, input().split())
pts = []
tree = node2d(0, XY_LIMIT-1)
for i in range(N):
    x, y, w = map(int, input().split())
    pts.append(point(x, y))
    tree.insert(x, y, w)

# check all pairs of points as lower boundaries
result = 0
for l in range(N):
    for b in range(N):
        value = tree.sum(pts[l].x, min(XY_LIMIT-1, pts[l].x + W - 1),
                         pts[b].y, min(XY_LIMIT-1, pts[b].y + H - 1))
        result = max(result, value)
print(result)

File: garbage/submissions/partially_accepted/garbage-sub_3_st.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

#define XY_LIMIT 1000000000

struct node {
    ll low, high;
    ll val = 0;
	node *left = nullptr, *right = nullptr;

	node(ll low, ll high)
        : low(low), high(high) {}

    void insert(ll pos, ll val) {
        this->val += val;
        if (low == high) return;

        ll mid = (low + high) / 2;
        if (pos <= mid) {
            if (!left) left = new node(low, mid);
            left->insert(pos, val);
        }
        else {
            if (!right) right = new node(mid+1, high);
            right->insert(pos, val);
        }
    }

    ll sum(ll l, ll h) {
        if (h < low || l > high) return 0;
        if (l <= low && h >= high) return val;

        ll res = 0;
        if (left) res += left->sum(l, h);
        if (right) res += right->sum(l, h);
        return res;
    }
};

struct node2d {
    ll low, high;
    node *tree;
	node2d *left = nullptr, *right = nullptr;

	node2d(ll low, ll high)
        : low(low), high(high), tree(new node(0, XY_LIMIT-1)) {}

    void insert(ll x, ll y, ll val) {
        tree->insert(y, val);
        if (low == high) return;

        ll mid = (low + high) / 2;
        if (x <= mid) {
            if (!left) left = new node2d(low, mid);
            left->insert(x, y, val);
        }
        else {
            if (!right) right = new node2d(mid+1, high);
            right->insert(x, y, val);
        }
    }

    ll sum(ll l, ll r, ll b, ll t) {
        if (r < low || l > high) return 0;
        if (l <= low && r >= high) return tree->sum(b, t);

        ll res = 0;
        if (left) res += left->sum(l, r, b, t);
        if (right) res += right->sum(l, r, b, t);
        return res;
    }
};

int main() {
    // read input
    int N, W, H;
    cin >> N >> W >> H;
    vector<pair<int, int>> pts;
    node2d *tree = new node2d(0, XY_LIMIT-1);
    for (int i = 0; i < N; i++) {
        int x, y, w;
        cin >> x >> y >> w;
        pts.push_back({x, y});
        tree->insert(x, y, w);
    }

    // check all pairs of points as lower boundaries
    ll result = 0;
    for (int l = 0; l < N; l++) {
        for (int b = 0; b < N; b++) {
            ll value = tree->sum(pts[l].first, min(XY_LIMIT-1, pts[l].first + W - 1),
                                 pts[b].second, min(XY_LIMIT-1, pts[b].second + H - 1));
            result = max(result, value);
        }
    }
    cout << result << endl;
}
